include/mscclpp/fifo_device.hpp

```diff
@@ -15,98 +15,89 @@
-/// The push() function increments head, hostTail is updated in Fifo::pop(), and it occasionally flushes
-/// it to tailReplica via Fifo::flushTail().
-///
-/// If `env()->fifoUseTailReplica` is false, FifoDeviceHandle::tailReplica points to the original tail on the host.
-/// In this case, the tail is always up-to-date and there is no need to flush it to the device.
-///
-/// Duplicating the tail is a good idea because the FIFO is large enough, and we do not need frequent updates for the
-/// tail as there is usually enough space for device threads to push their work into.
-///
+/// Concurrent FIFO for multiple device threads to push work elements and a single host proxy thread to consume them.
+/// Head pointer is on device, tail pointer is on host (readable by device).
 struct FifoDeviceHandle {
 #if defined(MSCCLPP_DEVICE_COMPILE)
   /// Push a trigger to the FIFO.
-  ///
-  /// @param trigger The trigger to push.
-  /// @param maxSpinCount The maximum number of spin counts before asserting. Never assert if negative.
-  /// @return The new head of the FIFO.
+  /// @param trigger Trigger to push.
+  /// @param maxSpinCount Max spin count before assert. Never assert if negative.
+  /// @return Previous head of the FIFO where the trigger was pushed.
   MSCCLPP_DEVICE_INLINE uint64_t push(ProxyTrigger trigger, [[maybe_unused]] int64_t maxSpinCount = 1000000) {
-    uint64_t curFifoHead = atomicFetchAdd(this->head, (uint64_t)1, memoryOrderRelaxed);
-
-    // make the last bit intentionally non-zero so that we can safely poll. Don't worry, we will change it back in host
-    // side
-    trigger.snd ^= ((uint64_t)1 << (uint64_t)63);
-
-    // Only one of two conditions need to be met to proceed. Either the tail has advanced enough or where we need to
-    // write to is 0. However, the first condition is faster to check since the tail is flushed periodically anyways but
-    // for the second condition we need to read CPU memory.
-    // As atomic access is slow, we first check using the bare pointer and then use the atomic load if the
-    // condition is not met.
-    if (curFifoHead >= size + *(this->tailReplica)) {
-      OR_POLL_MAYBE_JAILBREAK((curFifoHead >= size + atomicLoad(this->tailReplica, memoryOrderRelaxed)),
-                              (atomicLoad(&(this->triggers[curFifoHead % size].fst), memoryOrderRelaxed) != 0),
-                              maxSpinCount);
+    uint64_t prevHead = atomicFetchAdd(head, uint64_t{1}, memoryOrderRelaxed);
+
+    // Set last bit non-zero for safe polling; host will revert.
+    trigger.snd ^= (uint64_t{1} << uint64_t{63});
+
+    // Proceed if tail advanced or target slot is 0.
+    // As tail value is cached on device, the device doesn't need to access host memory every time.
+    uint64_t numInflights = prevHead - *tailCache;
+    if (numInflights >= size / 2) {
+      numInflights = prevHead - (*tailCache = uncachedLoad(tail));
+    }
```

Comment:
Do we really need this? I think the cost is almost same: Call load every size/2 elems vs call load every size elems. The POLL_MAYBE_JARIBREAK just a simple while loop and don't have too much overhead.

Tested in H200 and A100, the perf is almost similar.